backend-go/lib.go

```diff
@@ -0,0 +1,324 @@
+	}
+}
+
+func loadConfig() map[string]any {
+	config := map[string]any{
+		"CONFIG_FILE": nil,
+		"CONFIG_KEY":  nil,
+	}
+	updateConfig(config, parseConsoleArguments())
+	populateConfig(config, getEnvs())
+	if _, contained := config["DEV"]; !contained {
+		config["DEV"] = false
+	}
+
+	populateConfig(config, getDefaultConfig(config["DEV"].(bool)))
+	populateConfig(config, getFileConfig(config))
+	return config
+}
+
+func updateConfig[V any | string](config map[string]any, updates map[string]V) {
+	for k, v := range updates {
+		config[k] = v
+	}
+}
+
+func populateConfig[V any | string](config map[string]any, updates map[string]V) {
+	for k, v := range updates {
+		// check if key is already in map and populate if it isn't
+		if _, contained := config[k]; !contained {
+			config[k] = v
+		}
+	}
+}
+
+func getFileConfig(config map[string]any) map[string]any {
+	// Note that the file config is supposed to be a json config!
+	jsonConfig := map[string]any{}
+
+	configPath := config["CONFIG_FILE"].(string)
+	if len(configPath) == 0 {
+		return jsonConfig
+	}
+
+	jsonConfigBytes, err := os.ReadFile(configPath)
+
+	if err != nil {
+		log.Printf("Config file %s could not be read: %s\n", configPath, err)
+		return jsonConfig
+	}
+
+	err = json.Unmarshal(jsonConfigBytes, &jsonConfig)
+	if err != nil {
+		log.Printf("Config file %s could not be parsed: %s\n", configPath, err)
+		return jsonConfig
+	}
+
+	return jsonConfig
+}
+
+func getDefaultConfig(devFlag bool) map[string]string {
+	if !devFlag {
+		return map[string]string{
+			"NATS": "true", // at the moment only nats is supported
+		}
+	}
+
+	log.Println("Running in development mode. Using default values for missing environment variables.")
+	return map[string]string{
+		"NATS":         "true", // at the moment only nats is supported
+		"NATS_URL":     nats.DefaultURL,
+		"SERVICE_NAME": fmt.Sprint("dev-", os.Getgid()),
+		"DATA_DIR":     filepath.Join(".", "data"), // we use filepath here to get the correct separator
+	}
+}
+
+func getEnvs() map[string]string {
+	envs := map[string]string{}
+	for _, entry := range os.Environ() {
+		split := strings.Split(entry, "=")
+		key, val := split[0], split[1]
+
+		// we don't want to overwrite config entries with empty environment variables!
+		if len(val) == 0 {
+			continue
+		}
+
+		envs[key] = val
+	}
+	return envs
+}
+
+func parseConsoleArguments() map[string]any {
+	// setup flags
+	var (
+		devFlag  bool
+		natsFlag bool
+
+		natsUrlFlag  string
+		natsUserFlag string
+		natsPwdFlag  string
+
+		configFileFlag string
+		configKeyFlag  string
+
+		serviceNameFlag string
+		dataDirFlag     string
+	)
+
+	warnDescription := "(Note: It is recommended to set this via the environment variables or the config!)"
+
+	flag.BoolVar(&devFlag, "development mode", false, "If set, program will start in development mode")
+	flag.BoolVar(&natsFlag, "use NATS", true, "If set, service will use NATS as the messaging bus (default)")
```

Comment:
Feel free to leave it in if you like, but I don't think it's necessary to specify whether to use NATS from the outside. I.e., the service implementation should know (and specify, as a parameter to `StartService()`) whether it needs NATS.

I don't think it makes sense to
1. deny a service that uses NATS a NATS connection through the configuration, or
2. (from the outside) forcing a NATS connection when a service's implementation declares this to be unnecessary.

Reply:
Removed from cli config parser